Spawn Perma Textscreens after Map Cleanup

ChristianLW opened the pull request:
Make the Perma Textscreens appear again after cleaning up the map

Cherry commented:
Thanks for the PR!

If `game.CleanUpMap` is called with extra filters, such as `game.CleanUpMap(true, {"sammyservers_textscreen"})` won't this result in the textscreens being spawned twice? 

Perhaps we should add a check to see if the textscreen is already spawned?

ChristianLW commented:
That's true, and something I didn't think about.
It is something I'm unsure how to do, though, as neither PreCleanUpMap or PostCleanUpMap gives any information. Maybe just checking if any textscreens are there? I mean, the player wouldn't be able to spawn a textscreen, between the map being cleaned up and GM:PostCleanUpMap being called, right?

Cherry commented:
The best solution I imagine would be to list entities by class of `sammyservers_textscreen`, and if one already exists with the same `.uniqueName` when auto-spawning the textscreens, don't spawn it. There'll technically be a minimal chance for collision here with the `StringRandom` function I'm using, but I think it'll be fine.

ChristianLW commented:
So either test for every textscreen, whether it already exists, or create a copy of the textscreen table and remove any screens from the copy that already exist, then spawn the rest.

mcNuggets1 commented:
for k, v in ipairs(ents.FindByClass("3d2dtext")) do
     if v.uniqueName == tab.uniqueName then
             continue
     end
end

Cherry commented:
Something like what @mcNuggets1 would work as a test. I'd recommend storing the result once though, rather than iterating for every single textscreen. There's also no need to use `ipairs` here. There's a performance hit when using it, and `pairs` would work just fine.

Cherry commented:
> What are you talking about?
> ipairs is faster than pairs.

From my understanding, it's not. `pairs` is almost always faster other than on very tiny tables. I've had people show me textscreens files with thousands, or even over ten thousand items before.

https://forum.facepunch.com/gmoddev/mzmp/Speed-Analysis-of-pairs-and-ipairs/1/#postelftd
https://springrts.com/wiki/Lua_Performance#TEST_9:_for-loops

-----

Completely besides the point of this pull request though. Let's get back on topic please. Do you want to update your Pull Request with these changes @ChristianLW?

mcNuggets1 commented:
It's slightly faster.
The threads seem to be outdated.

https://i.imgur.com/U9dJCWT.png

Cherry commented:
Okay, thanks. We can optimise after the fact. Let's get back on topic now.

ChristianLW commented:
What do you mean by storing the result once?
Breaking out of the loop, if it finds a match?

Cherry commented:
Something like this. (Completed untested).
```lua
local function SpawnPermaTextscreens()
	print("Spawning textscreens...")
	textscreens = file.Read("sammyservers_textscreens.txt", "DATA")

	if not textscreens then textscreens = {} return	end

	textscreens = util.JSONToTable(textscreens)
	local existingTextscreens = {}
	for k,v in pairs(ents.FindByClass("sammyservers_textscreens")) do
		if not v.uniqueName then continue end
		existingTextscreens[v.uniqueName] = true
	end
	local count = 0

	for k, v in pairs(textscreens) do
		if v.MapName ~= game.GetMap() then continue end
		if existingTextscreens[v.uniqueName] then continue end

		local textScreen = ents.Create("sammyservers_textscreen")
		textScreen:SetPos(Vector(v.posx, v.posy, v.posz))
		textScreen:SetAngles(Angle(v.angp, v.angy, v.angr))
		textScreen.uniqueName = v.uniqueName
		textScreen:Spawn()
		textScreen:Activate()
		textScreen:SetMoveType(MOVETYPE_NONE)

		for k, v in pairs(v.lines or {}) do
			textScreen:SetLine(k, v.text, Color(v.color.r, v.color.g, v.color.b, v.color.a), v.size, v.font)
		end

		textScreen:SetIsPersisted(true)
		count = count + 1
	end

	print("Spawned " .. count .. " textscreens for map " .. game.GetMap())
end
```

ChristianLW commented:
Tested it, and it works (other than a simple entity class typo, that I fixed)
I've updated the thing. I don't know what the thing is called. I'm pretty new to Github.
I think it's the branch, that I updated, or the pull request, or both.
Anyway, tested and updated.

Cherry commented:
GitHub was having a lot of issues last night preventing me from merging this. I'll get it released today. Thanks again for the great PR! 😃

Cherry commented:
And released. https://steamcommunity.com/sharedfiles/filedetails//changelog/109643223/
